### PR TITLE
Adds the capability to exploit powershell remoting via HTTP

### PR DESCRIPTION
**Pull request by VakarisZ**
# What does this PR do? 

Adds the capability to exploit powershell remoting via HTTP

Add any further explanations here. 

## PR Checklist
* [ ] Have you added an explanation of what your changes do and why you'd like to include them?
* [ ] Is the TravisCI build passing? 
* [ ] Was the CHANGELOG.md updated to reflect the changes?
* [ ] Was the documentation framework updated to reflect the changes?

## Testing Checklist

* [ ] Added relevant unit tests?
* [ ] Have you successfully tested your changes locally? Elaborate:
    > Tested by running with credentials over HTTP
    > Running with credentials over HTTPS
* [ ] If applicable, add screenshots or log transcripts of the feature working

## Explain Changes

Are the commit messages enough? If not, elaborate.

### CODE REVIEW

File: monkey/infection_monkey/exploit/powershell.py [hunk 1]
```diff
@@ -41,49 +47,79 @@ def _set_sensitive_packages_log_level_to_error():
...
 
     def _exploit_host(self):
-        self.client = self._authenticate_via_brute_force()
+        is_https = self._is_client_using_https()
+        credentials = CredentialGenerator(
+            self.host.ip_addr,
+            self._config.exploit_user_list,
+            self._config.exploit_password_list,
+            is_windows_os(),
+        ).get_credentials(is_https=is_https)
```
Comment (mssalvatore) [inline monkey/infection_monkey/exploit/powershell.py:56]: This doesn't need to be a class. This is effectively just a complicated function invocation. The class doesn't actually store any state.

File: monkey/infection_monkey/exploit/powershell.py [hunk 2]
```diff
@@ -41,49 +47,79 @@ def _set_sensitive_packages_log_level_to_error():
...
-        credentials = utils.get_credentials(
-            self._config.exploit_user_list, self._config.exploit_password_list, is_windows_os()
+    def _is_client_using_https(self) -> bool:
+        try:
+            self._try_http()
+            return False
+        except AuthenticationError:
+            return False
+        except Exception:
+            pass
```
Comment (mssalvatore) [inline monkey/infection_monkey/exploit/powershell.py:71]: Maybe put debug logging here?

File: monkey/infection_monkey/exploit/powershell_utils/utils.py [hunk 3]
```diff
@@ -72,3 +19,38 @@ def build_monkey_execution_command(host: VictimHost, depth: int, executable_path
...
+
+def get_client_based_on_auth_options(auth_options: AuthOptions) -> IClient:
+
+    # Passwordless login only works with SSL false, AUTH_BASIC and ENCRYPTION_NEVER
+    if auth_options.password == "":
+        ssl = False
+    else:
+        ssl = auth_options.is_https
+    auth = AUTH_NEGOTIATE if auth_options.password != "" else AUTH_BASIC
+    encryption = ENCRYPTION_AUTO if auth_options.password != "" else ENCRYPTION_NEVER
```
Comment (mssalvatore) [inline monkey/infection_monkey/exploit/powershell_utils/utils.py:45]: This code has unit tests that now fail.

File: monkey/infection_monkey/exploit/powershell_utils/auth_options.py [hunk 4]
```diff
@@ -0,0 +1,10 @@
+from dataclasses import dataclass
+from typing import Union
+
+
+@dataclass
+class AuthOptions:
+    ip_addr: str
```
Comment (mssalvatore) [inline monkey/infection_monkey/exploit/powershell_utils/auth_options.py:7]: I don't think the IP address is really part of the authentication options. It's information about the host.

**Review by mssalvatore — CHANGES_REQUESTED** [review]

File: monkey/infection_monkey/exploit/powershell.py (hunk 1 above)
Comment (VakarisZ) [reply monkey/infection_monkey/exploit/powershell.py]: Agreed.

**Pull request merged**